File: scripts/add-lang-i18n.js
/**
 * Adds new language translations to i18n.js.
 * Uses English as fallback for new languages when no specific translation exists.
 * Run: node scripts/add-lang-i18n.js
 */
const fs = require('fs');
const path = require('path');

const NEW_LANGS = ['uk','ro','el','sk','he','fil','ms','bn'];

const i18nPath = path.join(__dirname, '../web/static/js/i18n.js');
let content = fs.readFileSync(i18nPath, 'utf8');

// Match each translation block: 'key': { ... }
// We need to add new languages before the closing };
const blockRegex = /'([^']+)':\s*\{([^}]+)\}/g;

content = content.replace(blockRegex, (match, key, inner) => {
  // Skip if already has uk (new langs already added)
  if (inner.includes("uk:'")) return match;

  // Extract en value
  const enMatch = inner.match(/en:'((?:[^'\\]|\\.)*)'/);
  const enVal = enMatch ? enMatch[1].replace(/\\/g, '\\\\').replace(/'/g, "\\'") : '';

  // Extract hi value (last lang before our addition)
  const hiMatch = inner.match(/hi:'((?:[^'\\]|\\.)*)'\s*$/);
  const hiVal = hiMatch ? hiMatch[1] : '';

  // Build additions for new languages - use en as fallback
  let add = '';
  for (const code of NEW_LANGS) {
    add += `, ${code}:'${enVal}'`;
  }

  // Replace hi:'xxx' at end with hi:'xxx', uk:'...', ...
  const newInner = inner.replace(/hi:'((?:[^'\\]|\\.)*)'\s*$/, (m) => m + add);
  return `'${key}': {${newInner}}`;
});

fs.writeFileSync(i18nPath, content);
console.log(`Added ${NEW_LANGS.join(', ')} to all translation keys. New languages use English as fallback.`);

File: scripts/add_lang_i18n.py
#!/usr/bin/env python3
"""Add new languages to all i18n keys. Uses English as fallback.
Run: python scripts/add_lang_i18n.py
"""
import re

# New languages: add after last existing (bn). Uses en as fallback for consistency.
NEW_LANGS = ['sr','bg','hr','sl','lt','lv','et','fa','sw','af','ca','gl','ta','te','ml','ur','az','hy','ka','si','ne','my','km']
I18N_PATH = 'web/static/js/i18n.js'

def process_line(line):
    # Skip if already has first new lang
    if any(f"{lang}:'" in line for lang in NEW_LANGS[:1]):
        return line
    if 'bn:' not in line or "': {" not in line:
        return line
    en_m = re.search(r"en:'((?:[^'\\]|\\.)*)'", line)
    if not en_m:
        return line
    en_val = en_m.group(1).replace("\\'", "'").replace("\\\\", "\\")
    en_escaped = en_val.replace("\\", "\\\\").replace("'", "\\'")
    additions = ''.join(f", {lang}:'{en_escaped}'" for lang in NEW_LANGS)
    new_line = re.sub(r"(bn:'(?:[^'\\]|\\.)*')\s*(\})", r"\1" + additions + r" \2", line)
    return new_line

def main():
    with open(I18N_PATH, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    out = []
    changed = 0
    for line in lines:
        new_line = process_line(line)
        if new_line != line:
            changed += 1
        out.append(new_line)
    with open(I18N_PATH, 'w', encoding='utf-8') as f:
        f.writelines(out)
    print(f"Added {len(NEW_LANGS)} languages to {changed} keys (English fallback).")

if __name__ == '__main__':
    main()
